### DISCUSSION TITLE
MPFS MSS USB UVC Support?

### DISCUSSION THREAD

**Discussion by mcmordie** [polarfire-soc/polarfire-soc-documentation#119] [Applications and demos] [answered]
The PolarFire SoC includes an Microprocessor Subsystem (MSS) USB 2.0 On The Go (OTG) peripheral which can be assigned to an Ultra Low Pin Count (ULPI) interface on IO Bank 2 (per [the reference manual](https://onlinedocs.microchip.com/pr/GUID-0E320577-28E6-4365-9BB8-9E1416A0A6E4-en-US-3/index.html?GUID-E7EAC8F8-2757-4881-BD09-325207D9626D)) .

I would like to connect this via a USB3340 ULPI transceiver and use it as a slave Universal Video Class (UVC) capture device.  

Questions:

1. Is this feasible / sensible?
2. What software support is available under Linux to make this happen?

Thanks in advance.

Dave

**Comment by hughbreslin** [comment]
Hi @mcmordie 

First thing, are you trying this on an Icicle Kit? The kit itself has a USB3340 connected up to the MSS ULPI interface so this might be a good place to test it out. 

The Icicle Kit seems to be demonstrating what you're looking for in terms of hardware. 

From the software side we have used USB cameras to the microUSB connector on the kit, we haven't tried UVC support in Linux, we know it is available to be used in a kernel configuration but you'll have to enable it yourself in a custom build. We don't have a definitive list of all of the config options you'll need to enable for this unfortunately.

**Reply by mcmordie** [reply]
@hughbreslin just to make sure we are understand each other, I want the Linux device to act as a UVC Device, so that an external host essentially sees the MSS as a webcam-- this is sometimes known as UVC gadget support.  Can you confirm there is support for this in the Kernel?

I do have an Icicle eval kit and for some reason had not noticed that the connection is exactly the same as what I want to do.  So thanks for that, I will seek out the UVC gadget support and if I can find it, enable and test.  Not at all clear how this works from user space, but I guess it must be through V4L2.

**Reply by mcmordie** [reply]
Did some digging and indeed @hughbreslin is right, there is UVC "gadget" support in the kernel, just have to turn it on and it should work on the icicle dev kit.  Not quite done getting the configuration set up, but I found this resource helpful.  Thanks for your help @hughbreslin!

https://developer.ridgerun.com/wiki/index.php?title=How_to_use_the_UVC_gadget_driver_in_Linux

**Reply by mcmordie** [reply]
These instructions break down at UDC device support, which seems not to be enabled by default.  Any advice for how to get the USB3380 supported in peripheral mode via ULPI in the kernel?  I see just the following in the device tree:

```
&usb {
	status = "okay";
	dr_mode = "otg";
};
```

From the driver standpoint we need UDC support to be enabled, not sure what changes are needed to get this.

**Reply by mcmordie** [reply]
Update: have changed "otg" to peripheral in the dts-- this did not help.  When I ls /sys/class/udc after running my bind script, the dir is empty.  message in dmesg is:

` 389.878053] udc-core: couldn't find an available UDC - added [g_webcam] to list of pending drivers`

`lsmod` gives me hope that I am almost there:

```
Module                  Size  Used by
usb_f_uvc              86016  6
videobuf2_vmalloc      16384  1 usb_f_uvc
videobuf2_dma_sg       20480  1 usb_f_uvc
videobuf2_memops       16384  2 videobuf2_vmalloc,videobuf2_dma_sg
videobuf2_v4l2         36864  1 usb_f_uvc
videobuf2_common       81920  5 usb_f_uvc,videobuf2_vmalloc,videobuf2_v4l2,videobuf2_dma_sg,videobuf2_memops
videodev              352256  3 usb_f_uvc,videobuf2_v4l2,videobuf2_common
mc                     57344  3 videodev,videobuf2_v4l2,videobuf2_common
g_webcam               16384  0
libcomposite           81920  15 usb_f_uvc,g_webcam
udc_core               36864  2 usb_f_uvc,libcomposite
mpfs                   16384  0
phy_generic            20480  2 mpfs
pac193x                28672  0
industrialio          118784  1 pac193x
```

 Have attached the patches I've applied to get to here.  Would be grateful for any guidance anyone can provide!

Bind script:

```
#!/bin/bash
set -e

modprobe g_webcam

UDC=`ls /sys/class/udc`
UDC_ROLE=/sys/devices/platform/soc/78d9000.usb/ci_hdrc.0/role
CONFIGFS="/sys/kernel/config"
GADGET="$CONFIGFS/usb_gadget"
VID="0x0525"
PID="0x0102"
SERIAL="0123456789"
MANUF=$(hostname)
PRODUCT="UVC Gadget"1~UDC=`ls /sys/class/udc`
UDC_ROLE=/sys/devices/platform/soc/78d9000.usb/ci_hdrc.0/role
CONFIGFS="/sys/kernel/config"
GADGET="$CONFIGFS/usb_gadget"
VID="0x0525"
PID="0x0102"
SERIAL="0123456789"
MANUF=$(hostname)
PRODUCT="UVC Gadget"

mkdir -p $GADGET/g1

cd $GADGET/g1
echo $VID > idVendor
echo $PID > idProduct

mkdir -p strings/0x409

echo $SERIAL > strings/0x409/serialnumber
echo $MANUF > strings/0x409/manufacturer
echo $PRODUCT > strings/0x409/product

mkdir -p configs/c.1
mkdir -p configs/c.1/strings/0x409

CONFIG="configs/c.1"
FUNCTION="uvc.0"
mkdir -p functions/$FUNCTION

mkdir -p functions/$FUNCTION/streaming/uncompressed/u/360p

ls functions/uvc.0/streaming/
ls functions/uvc.0/streaming/uncompressed/u/
ls functions/uvc.0/streaming/uncompressed/u/360p/

cat <<EOF > functions/$FUNCTION/streaming/uncompressed/u/360p/dwFrameInterval
666666
1000000
5000000
EOF

cat functions/uvc.0/streaming/uncompressed/u/360p/dwFrameInterval

mkdir -p functions/$FUNCTION/streaming/header/h
cd functions/$FUNCTION/streaming/header/h
ln -s ../../uncompressed/u
cd ../../class/fs
ln -s ../../header/h
cd ../../class/hs
ln -s ../../header/h
cd ../../class/ss
ln -s ../../header/h
cd ../../../control
mkdir -p header/h
ln -s header/h class/fs
ln -s header/h class/ss
cd ../../../

echo 2048 > functions/$FUNCTION/streaming_maxpacket

ln -s functions/$FUNCTION configs/c.1

ls -l

echo $UDC > UDC

echo peripheral > $UDC_ROLE
```

[yocto-kernel-mods-polarfire-uvc-gadget.zip](https://github.com/polarfire-soc/polarfire-soc-documentation/files/9074968/yocto-kernel-mods-polarfire-uvc-gadget.zip)

**Reply by hughbreslin** [reply]
Hi @mcmordie unfortunately I don't have a huge amount to add in, sorry about the confusion re the host / device mode initially, this isn't something we've gone near yet! One thing to check on your set up is the board config, after a bit of checking into this the default configuration is J15 closed from the board schematic which would tie the USB_ID signal to ground:

![image](https://user-images.githubusercontent.com/44862881/178305614-30ecc950-3d48-4231-82c0-8833e37d672d.png)

My understanding is:

    USB HOST role
        The USBID pin (microcontroller): unconnected
        ID pin (USB connector): connected to ground
    USB OTG role
        The USBID pin from microcontroller: connected to ID pin (USB connector)
    USB DEVICE role
        USBID pin (microcontroller): connected to ID pin (USB connector)
        or
        USBID pin (microcontroller): unconnected and ID pin (USB connector): unconnected

**Reply by mcmordie** [reply]
Thanks for this thought-- JP15 was already open on my board.

**Comment by mcmordie** [comment, marked as answer]
Thanks for your help & thoughts @hughbreslin.  In the end, it was a matter of disabling USB host mode in the kernel and setting the USB driver dtc node to peripheral mode.

![image](https://user-images.githubusercontent.com/5338911/178360677-b24a379d-3220-48cb-aec5-e6d78ff39b58.png)

I have attached the yocto changes needed-- looks like lots of duplication in my cfg file (beware).  The key point after adding the gadget drivers was disabling USB host mode as depicted below:

![image](https://user-images.githubusercontent.com/5338911/178366608-c573eda2-a84f-45de-ad16-7cbd14eef538.png)

[uvc-kernel-mods.zip](https://github.com/polarfire-soc/polarfire-soc-documentation/files/9087770/uvc-kernel-mods.zip)

**Reply by hughbreslin** [reply]
Woohoo glad you got it working 🙂 thanks for sharing the steps
